internal/ingest/ingest.go

```diff
@@ -202,19 +277,53 @@ func (ing *Ingester) Sync(ctx context.Context, peerID peer.ID, peerAddr multiadd
 		//
 		// Reference to the latest synced CID is only updated if the given
 		// selector is nil.
-		c, err := ing.sub.Sync(ctx, peerID, cid.Undef, sel, peerAddr)
+		var seenAdCids []cid.Cid
+		c, err := ing.sub.SyncWithHook(ctx, peerID, cid.Undef, sel, peerAddr, func(i peer.ID, c cid.Cid) {
+			seenAdCids = append(seenAdCids, c)
+		})
+		// TODO(mm) we should have a mechanism that lets a manual sync always update the latest head in go legs.
+		if sel != nil && len(seenAdCids) > 0 {
+			event := legs.SyncFinished{
+				Cid:        seenAdCids[0],
+				PeerID:     peerID,
+				SyncedCids: seenAdCids,
+			}
+			ing.runIngestStep(event)
+		}
 		if err != nil {
 			log.Errorw("Failed to sync with provider", "err", err)
 			return
 		}
 		// Do not persist the latest sync here, because that is done in
 		// after we've processed the ad.
 
-		ing.signalMetricsUpdate()
-		// Notification channel; buffered so as not to block if no reader.
-		out <- c.Hash()
-	}()
+		// If latest head had already finished syncing, then do not wait for
+		// syncDone since it will never happen.
+		if latest == c {
+			log.Infow("Latest advertisement already processed", "adCid", c)
+			out <- c
+			return
+		}
```

Comment:
This may be wrong if a custom depth is given.  Does the wait break? Can we wait on something else?